app/javascript/hyacinth_v1/components/shared/forms/buttons/ProgressButton.js

```diff
@@ -0,0 +1,93 @@
+ *                                   will affect the state of this component.
+ * @param {function} props.onSuccess - Function to call when going back to the default state after success.
+ *                                     Function is given the result of the promise returned by the onClick function;
+ * @param {function} props.onError - Function to call when going back to the default state after an error.
+ *                                   Function is given the error from the promise returned by the onClick function;
+ */
+const ProgressButton = ({
+  label, type, loadingLabel, onClick, onSuccess, onError,
+}) => {
+  const [state, setState] = useState('default');
+  const successStateReturnTime = 500; // milliseconds
+  const errorStateReturnTime = 1500; // milliseconds
+  const minimumLoadingStateTime = 500; // milliseconds
+
+  let currentLabel = label;
+  let currentVariant = 'primary';
+
+  if (state === 'loading') {
+    currentLabel = loadingLabel;
+  } else if (state === 'success') {
+    currentLabel = 'Success!';
+    currentVariant = 'success';
+  } else if (state === 'error') {
+    currentLabel = 'An error occurred.';
+    currentVariant = 'danger';
+  }
+
+  const returnToDefaultStateAfterTimeout = (callback, returnTime) => {
+    setTimeout(() => {
+      setState('default');
+      callback();
+    }, returnTime);
+  };
+
+  const onClickHandler = (e) => {
+    e.preventDefault();
+    const expectedPromise = onClick();
+
+    if (!(expectedPromise instanceof Promise)) {
+      throw new Error('Expected onClick return value to be a promise.');
+    }
+
+    setState('loading');
+
+    Promise.all([
+      expectedPromise,
+      // Below, we're forcing a minimum time display of the loading state so that the loading
+      // message doesn't fly by in a jarring way if the promise resolves too quickly.
+      new Promise((resolve) => { setTimeout(resolve, minimumLoadingStateTime); }),
+    ])
+      // we only want to pass along the expectedPromise result
+      .then(allResult => allResult[0]).then((result) => {
+        setState('success');
+        returnToDefaultStateAfterTimeout(() => { onSuccess(result); }, successStateReturnTime);
+      }).catch((error) => {
+        setState('error');
+        returnToDefaultStateAfterTimeout(() => { onError(error); }, errorStateReturnTime);
+      });
+  };
+
+  return (
+    <Button type={type} variant={currentVariant} disabled={state === 'loading'} onClick={onClickHandler}>
+      {currentLabel}
+    </Button>
+  );
+};
+
+ProgressButton.propTypes = {
+  label: PropTypes.string.isRequired,
+  type: PropTypes.string.isRequired,
+  loadingLabel: PropTypes.string.isRequired,
+  onClick: PropTypes.func.isRequired,
+  onSuccess: PropTypes.func,
+  onError: PropTypes.func,
+};
+
+ProgressButton.defaultProps = {
+  onSuccess: () => {},
+  onError: () => {},
+};
+
+export default ProgressButton;
```

Comment:
great to be seeing 'success' now, instead of that misleading yellow 'could not save', on the dynamic field form as well!

Reply:
Yes!  I'm so glad that's gone now!